financial-templates-lib/price-feed/BalancerPriceFeed.js

```diff
@@ -0,0 +1,78 @@
+const assert = require("assert");
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const { BlockHistory, PriceHistory } = require("./utils");
+
+// Gets balancer prices as well as historical prices
+class BalancerPriceFeed extends PriceFeedInterface {
+  constructor(logger, web3, getTime, abi, address, tokenIn, tokenOut, lookback) {
+    assert(tokenIn, "BalancerPriceFeed requires tokenIn");
+    assert(tokenOut, "BalancerPriceFeed requires tokenOut");
+    assert(lookback, "BalancerPriceFeed requires lookback");
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+    this.getTime = getTime;
+
+    this.contract = new web3.eth.Contract(abi, address);
+    this.currentPrice = null;
+    this.lastUpdateTime = null;
+    this.tokenIn = tokenIn;
+    this.tokenOut = tokenOut;
+    this.lookback = lookback;
+    this.blockHistory = BlockHistory(web3);
+
+    // Add a callback to get price, if an error is thrown returns undefined
+    this.priceHistory = PriceHistory(async number => {
+      try {
+        const result = await this.contract.methods.getSpotPriceSansFee(this.tokenIn, this.tokenOut).call(number);
+        return result;
+      } catch (err) {
+        this.logger.warn({
+          at: "BalancerPriceFeed.getPrice",
+          message: err.message,
+          number
+        });
```

Comment:
If this error throws and gets caught, will that corrupt the state of the PriceHistory object?

Reply:
it will return an undefined price at that timestamp, but bubbling this error up will cause the bot to crash currently. i figured since the higher level logic was ok with undefined prices, that this would be ok as well. what do you think should happen?